cms/server/ContestWebServer.py

```diff
@@ -160,20 +166,47 @@ def get_current_user(self):
             self.clear_cookie("login")
             return None
 
-        # Load the user from DB.
+        # Load user from DB.
         user = self.sql_session.query(User)\
-            .filter(User.contest == self.contest)\
-            .filter(User.username == username).first()
+            .filter(User.username == username)\
+            .first()
 
-        # Check if user exists and is allowed to login.
-        if user is None or user.password != password:
+        # Check if user exists.
+        if user is None:
             self.clear_cookie("login")
             return None
-        if config.ip_lock and user.ip is not None \
-                and not check_ip(self.request.remote_ip, user.ip):
+
+        # Load participation from DB.
+        participation = self.sql_session.query(Participation)\
+            .filter(Participation.contest == self.contest)\
+            .filter(Participation.user == user)\
+            .first()
+
+        # Check if participaton exists.
+        if participation is None:
             self.clear_cookie("login")
             return None
-        if config.block_hidden_users and user.hidden:
+
+        # If a contest-specific password is defined, use that. If it's
+        # not, use the user's main password.
```

Comment:
Recentyl I thought about that and my stance now is this: we should keep only `Participation.password` *and* we should move `User.username` to `Participation.username` (so a user can be authenticated as `f.lastname` in one contest and as `f.lastname2` in another contest where there is a participating user with the same first name and last name).

Furthermore, since that would mean that username/password is *only* an authentication mechanism (and a rather "old" one in the OAuth2 era...) my proposal is to ditch: `User.username`, `User.password`, `Participation.password`, and add: `Participation.credentials` as a JSON string which depends on the authentication mechanism decided by the admins of that particular contest.